File: Day 15 - Stack and Queue/Next Greater Element II.cpp
class Solution
{
public:
  vector<int> nextGreaterElements(vector<int> &nums)
  {
    int n = nums.size();
    stack<int> s;
    vector<int> res(n, -1);

    for (int i = 2 * n - 1; i >= 0; --i)
    {
      int num = nums[i % n];
      while (!s.empty() && s.top() <= num)
        s.pop();
      if (!s.empty() && i < n)
        res[i] = s.top();
      s.push(num);
    }
    return res;
  }
};

File: Day 15 - Stack and Queue/Next Greater Element II.java
public class Solution {
    public int[] nextGreater(int[] A) {
        Stack<Integer> st = new Stack<>();
        int n = A.length;
        int[] ans = new int[n];
        for (int i = n - 1; i >= 0; i--) {
            int curr = A[i];
            while (!st.isEmpty() && st.peek() <= curr)
                st.pop();
            if (st.isEmpty())
                ans[i] = -1;
            else
                ans[i] = st.peek();
            st.push(curr);
        }
        return ans;
    }
}